File: Prácticas/deepspace/src/deepspace/PowerEfficientSpaceStation.java
package deepspace;

/**
 * Power efficient space station: gives more shot power and protection than a
 * usual space station.
 * 
 * @author Miguel Ángel Fernández Gutiérrez
 * @author Sergio Quijano Rey
 */
public class PowerEfficientSpaceStation extends SpaceStation {
    /**
     * Efficiency factor for power efficient space station.
     */
    private static final float EFFICIENCYFACTOR = 1.1f;
    
    /**
     * Class initializer.
     * @param station basic space station to convert to power efficient
     */
    PowerEfficientSpaceStation(SpaceStation station) {
        super(station);
    }
    
    /**
     * Make a shot.
     * @return the shot power
     */
    @Override
    public float fire() {
        return super.fire()*EFFICIENCYFACTOR;
    }
    
    /**
     * Use protection shield.
     * @return the shield's energy
     */
    @Override
    public float protection() {
        return super.protection()*EFFICIENCYFACTOR;
    }
    
    /**
     * Receives a loot.
     * @param loot loot to be received
     */
    @Override
    public Transformation setLoot(Loot loot) {
        Transformation trans = super.setLoot(loot);
        
        if ( trans == Transformation.SPACECITY )
            return Transformation.NOTRANSFORM;
        else
            return trans;
    }
    
    /**
     * To UI.
     */
    @Override
    public PowerEfficientSpaceStationToUI getUIversion() {
        return new PowerEfficientSpaceStationToUI(this);
    }
}

File: Prácticas/Deepspace/lib/PowerEfficientSpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStation'
require_relative 'Transformation'
require_relative 'PowerEfficientSpaceStationToUI'
#require_relative 'SuppliesPackage'


module Deepspace

# Power efficient space station: gives more shot power and protection than a usual space station
#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class PowerEfficientSpaceStation < SpaceStation

    # @!attribute [Float] efficiency factor for power efficient space station
    @@EFFICIENCYFACTOR = 1.10

    # Class initializer
    # @param station [SpaceStation] basic space station to convert to efficient
    # --Overriden
    def initialize(station)
        copy(station)
    end
    
    # Makes a shot
    # @return [Float] the shot power
    # --Overriden
    def fire
        return super*@@EFFICIENCYFACTOR
        # WIP COMPROBAR QUE ES @@EFFICIENCYFACTOR DE ESTA CLASE
    end

    # Use protection shield
    # @return [Float] the shield's energy
    # --Overriden
    def protection
        return super*@@EFFICIENCYFACTOR
    end

    # Receives a loot
    # @param [Loot] loot to be received
    # @return [Transformation] if loot received was efficient (GETEFFICIENT)
    #         or not (NOTRANSFORM)
    # --Overriden
    def setLoot(loot)
        super
        
        if loot.efficient
            return Transformation::GETEFFICIENT
        else
            return Transformation::NOTRANSFORM
        end
    end

end # class PowerEfficientSpaceStation

end # module Deepspace